network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+                        // TODO:  there are not separate keys for the full_node.  Are they therefore the same?
+                        // Note that signing_key type is Ed25519PublicKey
+                        // public key type is x25519::PublicKey
+                        signing_public_key: full_node
+                            .config()
+                            .full_node_network_identity_public_key
+                            .clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+
+    updates
+}
+
+impl ConfigurationChangeListener {
+    pub fn new(conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>, role: RoleType) -> Self {
+        Self {
+            conn_mgr_reqs_tx,
+            role,
+        }
+    }
+
+    /// Processes a received OnChainConfigPayload.  Depending on role (Validator or FullNode), parses
+    /// the appropriate configuration changes and passes it to the ConnectionManager channel.
+    async fn process_payload(&mut self, payload: OnChainConfigPayload) {
+        let node_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let updates = match self.role {
+            RoleType::Validator => extract_validator_updates(node_set),
+            RoleType::FullNode => extract_full_node_updates(node_set),
+        };
+
+        info!(
+            "Update {} Network about new Node IDs",
+            self.role.to_string()
+        );
+        for mut update in updates {
+            let err_msg;
+            match &update {
+                ConnectivityRequest::UpdateAddresses(a, b, c) => {
+                    err_msg = format!("Unable to update network address for peer {}", b);
+                    info!(
+                        "Updating {} network address for {} to {}",
+                        self.role.to_string(),
+                        b.to_string(),
+                        c[0].to_string()
+                    );
+                }
+                ConnectivityRequest::UpdateEligibleNodes(a) => {
+                    err_msg = format!(
+                        "Unable to update {} network's eligible peers",
+                        self.role.to_string()
+                    );
+                    info!(
+                        "Update {} network about new Node IDs",
+                        self.role.to_string()
+                    );
+                }
+                ConnectivityRequest::GetDialQueueSize(a) => {
+                    unimplemented!("Not expecting to receive this change!");
+                }
+            };
+            self.conn_mgr_reqs_tx.send(update).await.expect(&err_msg);
+        }
+    }
+
+    /// Starts the listener to wait on reconfiguration events.  Creates an infinite loop.
+    pub async fn start(
+        mut self,
+        mut reconfig_events: libra_channel::Receiver<(), OnChainConfigPayload>,
+    ) {
+        loop {
```

Comment:
if you wanted to venture into an integration test, which I really don't care if you do, you could consider splitting this out and loop around a helper function that actually does this process.